tests/phpunit/tests/dependencies/scripts.php

```diff
@@ -3065,4 +3065,171 @@ protected function assertEqualMarkup( $expected, $actual, $message = '' ) {
 	protected function add_html5_script_theme_support() {
 		add_theme_support( 'html5', array( 'script' ) );
 	}
+
+	/**
+	 * Test that a script is moved to the footer if it is made non deferrable, was in the header and
+	 * all scripts that depend on it are in the footer.
+	 *
+	 * @ticket 58599
+	 *
+	 * @dataProvider data_provider_script_move_to_footer
+	 *
+	 * @param string $b_in_footer in_footer value for script B.
+	 * @param string $expected    Expected output.
+	 * @param bool   $expect_a_in_footer Is script A in the footer?
+	 * @param bool   $expect_b_in_footer Is script B in the footer?
+	 */
+	public function test_wp_scripts_move_to_footer( $b_in_footer, $expected_header, $expected_footer, $expect_a_in_footer, $expect_b_in_footer, $c_in_footer, $d_in_footer ) {
+		// Script A uses strategy=>defer.
+		wp_enqueue_script( 'script-a', 'script-a.js', array(), null, array( 'strategy' => 'defer' ) );
+
+		// Script B depends on script A and uses variations on in_footer.
+		wp_enqueue_script( 'script-b', 'script-b.js', array( 'script-a' ), null, $b_in_footer );
+
+		// Script C depends also on script A and uses variations on in_footer.
+		wp_enqueue_script( 'script-c', 'script-c.js', array( 'script-a' ), null, $c_in_footer );
+
+		// Script D depends also on script B and uses variations on in_footer.
+		wp_enqueue_script( 'script-d', 'script-d.js', array( 'script-b' ), null, $d_in_footer );
+
+		// Run the header and footer capturing the output.
+		ob_start();
+		wp_scripts()->do_head_items();
+		$header = ob_get_clean();
+		ob_start();
+		wp_scripts()->do_footer_items();
+		$footer = ob_get_clean();
+
+		$this->assertEquals( $expected_header, $header );
+		$this->assertEquals( $expected_footer, $footer );
+		$this->assertEquals( $expect_a_in_footer, in_array( 'script-a', wp_scripts()->in_footer, true ) );
+		$this->assertEquals( $expect_b_in_footer, in_array( 'script-b', wp_scripts()->in_footer, true ) );
+	}
+
+	/**
+	 * Data provider for test_wp_scripts_move_to_footer.
+	 *
+	 * @return array[] {
+	 *     Array of arguments for test.
+	 *
+	 *     @type string $b_in_footer in_footer value for script B.
+	 *     @type string $expected    Expected output.
```

Comment:
There are args missing here. Maybe best to not duplicate what is indicated in `test_wp_scripts_move_to_footer`?